xds/internal/xdsclient/e2e_test/lds_watchers_test.go

```diff
@@ -0,0 +1,797 @@
+// invocation of all watch callbacks.
+//
+// The test is run with both old and new style names.
+func (s) TestLDSWatch_ThreeWatchesForDifferentResourceNames(t *testing.T) {
+	overrideFedEnvVar(t)
+	mgmtServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, nil)
+	defer cleanup()
+
+	// Create an xDS client with the above bootstrap contents.
+	client, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer client.Close()
+
+	// Register two watches for the same listener resource and have the
+	// callbacks push the received updates on to a channel.
+	updateCh1 := testutils.NewChannel()
+	ldsCancel1 := client.WatchListener(ldsName, func(u xdsresource.ListenerUpdate, err error) {
+		updateCh1.Send(xdsresource.ListenerUpdateErrTuple{Update: u, Err: err})
+	})
+	defer ldsCancel1()
+	updateCh2 := testutils.NewChannel()
+	ldsCancel2 := client.WatchListener(ldsName, func(u xdsresource.ListenerUpdate, err error) {
+		updateCh2.Send(xdsresource.ListenerUpdateErrTuple{Update: u, Err: err})
+	})
+	defer ldsCancel2()
+
+	// Register the third watch for a different listener resource.
+	updateCh3 := testutils.NewChannel()
+	ldsCancel3 := client.WatchListener(ldsNameNewStyle, func(u xdsresource.ListenerUpdate, err error) {
+		updateCh3.Send(xdsresource.ListenerUpdateErrTuple{Update: u, Err: err})
+	})
+	defer ldsCancel3()
+
+	// Configure the management server to return two listener resources,
+	// corresponding to the registered watches.
+	resources := e2e.UpdateOptions{
+		NodeID: nodeID,
+		Listeners: []*v3listenerpb.Listener{
+			e2e.DefaultClientListener(ldsName, rdsName),
+			e2e.DefaultClientListener(ldsNameNewStyle, rdsName),
+		},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Verify the contents of the received update for the all watchers. The two
+	// resources returned differ only in the resource name. Therefore the
+	// expected update is the same for all the watchers.
+	wantUpdate := xdsresource.ListenerUpdateErrTuple{
+		Update: xdsresource.ListenerUpdate{
+			RouteConfigName: rdsName,
+			HTTPFilters:     []xdsresource.HTTPFilter{{Name: "router"}},
+		},
+	}
+	if err := verifyListenerUpdate(ctx, updateCh1, wantUpdate); err != nil {
+		t.Fatal(err)
+	}
+	if err := verifyListenerUpdate(ctx, updateCh2, wantUpdate); err != nil {
+		t.Fatal(err)
+	}
+	if err := verifyListenerUpdate(ctx, updateCh3, wantUpdate); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestLDSWatch_ResourceCaching covers the case where a watch is registered for
+// a resource which is already in the cache.  The test verifies that the new
+// watch is handled from the cache, without a request being sent to the
```

Comment:
Super nit: handled from cache meaning? handle could mean a lot of things, including invoking watch callback with error. Can you switch this statement to receives update from resources present in cache or something.

Reply:
Done. Thanks.